[WIP] [DO NOT MERGE] Common reconciler plumbing code for eventing

akashrv opened the pull request:
This is a temporary PR to discuss design. Actual PR will be submitted to KNative eventing master.
What is this PR about?
 - We (Adam, Grant and Myself)  noticed while writing controllers that there is lot of plumbing code that is copy pasted and along with it certain bugs were copied over to all controllers.
- To fix this forever, I create a Reconciler that embeds the concrete reconcilers. This Reconciler has all the plumbing code and manages:
  - How finalizers are added correctly
  - How Deletion is handled correctly
  - How status update is done correctly
  - Consistent log messages

What is this PR not about?
- This is not an attempt to have a generic Reconciler. If that is the need to refer to KubeBuilder or Operator Framework. This is just to standardize the pattern inside KNative eventing.

Note:
 - I haven't run UTs and I am sure they will fail
 - I haven't run ITs
 - However I have run the controllers in by dev cluster and they work.
- I have modified subscription controller and GCPPudSub controller to this new pattern so that you can understand how less code we will have to write for new controllers. 

Once we feel the design is good enough, then I will work on the UTs, ITs, TODOs, changing all other controllers and final polishing.

Please feel free to share your feedback

evankanderson committed:
Remove markdown lint presubmit check in favor of the sockpuppet. (#907)

evankanderson committed:
Record the decision on Broker and Trigger (#862)

* Initial definitions of Broker and Trigger based on decision docs

Signed-off-by: Evan Anderson <evan.k.anderson@gmail.com>

* Interim commit to proxy content home.

* Address @vaikas-google comments.

* Update with @katfang wording

akashrv committed:
WIP. Common reconciler

akashrv committed:
Moved to a builder pattern instead of interface conversions

akashrv commented on pkg/reconciler/v1alpha1/broker/broker.go:
```diff
@@ -151,7 +151,7 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
 		r.recorder.Event(broker, corev1.EventTypeNormal, brokerReconciled, "Broker reconciled")
```
Please ignore broker.go. This file slipped in due to some other change I had in it

akashrv commented on pkg/reconciler/v1alpha1/broker/broker_test.go:
```diff
@@ -461,41 +461,6 @@ func TestReconcile(t *testing.T) {
 			},
```
Please ignore broker_test.go. This file slipped in due to some other change I had in it

n3wscott commented:
**I strongly recommend not doing this work.** I think if energy is put into reconciling logic improvements, it needs to be in `knative/pkg`. Furthermore, eventing and serving need to match in core implementation so that improvements like this can benefit Knative.

evankanderson commented on contrib/gcppubsub/pkg/controller/channel/controller.go:
```diff
@@ -38,46 +43,60 @@ const (
 			defaultSecretKey:    defaultSecretKey,
 			pubSubClientCreator: pubsubutil.GcpPubSubClientCreator,
 		}
+		builder := eventingreconciler.NewBuilder(rec).
```
Golang doesn't tend to use a lot of builders; instead, use varags with an "options" types, or even functional options.

https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

akashrv replied:
Thanks for the pointer. Interesting pattern, and solves exactly what I am doing with the builder. Will adopt to this pattern.

evankanderson commented on contrib/gcppubsub/pkg/controller/channel/controller.go:
```diff
@@ -38,46 +43,60 @@ const (
+			WithFilter(shouldReconcile).
+			WithLogger(logger).
+			WithRecorder(mgr.GetRecorder(controllerAgentName)).
+			WithFinalizer(finalizerName, rec.Finalize).
```
It's strange to need to pass `rec` again here.

evankanderson replied:
Maybe `rec` could implement a "Finalizable" interface?

evankanderson commented on contrib/gcppubsub/pkg/controller/channel/controller.go:
```diff
@@ -38,46 +43,60 @@ const (
+			WithLogger(logger).
+			WithRecorder(mgr.GetRecorder(controllerAgentName)).
+			WithFinalizer(finalizerName, rec.Finalize).
+			WithInjectClientFunc(rec.InjectClient)
```
Ditto, this seems like `rec` should just implement a `Injectable` interface.

evankanderson commented on contrib/gcppubsub/pkg/controller/channel/controller.go:
```diff
@@ -38,46 +43,60 @@ const (
+			WithRecorder(mgr.GetRecorder(controllerAgentName)).
+			WithFinalizer(finalizerName, rec.Finalize).
+			WithInjectClientFunc(rec.InjectClient)
+		r := builder.Build()
```
The other disadvantage of the Builder pattern in golang is that I have no idea what type `r` is. (It's the return type from `.Build()`, which apparently does no argument validation, because it doesn't return err.)

akashrv replied:
Agree

evankanderson commented on contrib/gcppubsub/pkg/controller/channel/controller.go:
```diff
@@ -38,46 +43,60 @@ const (
+			&source.Kind{
+				Type: &corev1.Service{},
+			},
+			&handler.EnqueueRequestForOwner{OwnerType: &eventingv1alpha1.Channel{}, IsController: true})
```
Was this auto-formatted by something? If so, was it something other than `go fmt`?

evankanderson commented on contrib/gcppubsub/pkg/controller/channel/controller.go:
```diff
@@ -86,3 +105,13 @@ func ProvideController(defaultGcpProject string, defaultSecret *corev1.ObjectRef
+// shouldReconcile determines if this Controller should control (and therefore reconcile) a given
+// Channel. This Controller only handles gcp-pubsub channels.
+func shouldReconcile(ctx context.Context, obj eventingreconciler.ReconciledResource, r record.EventRecorder) bool {
+	c := obj.(*eventingv1alpha1.Channel) // TODO: Is there a way to do this like try catch - to catch panicking?
```
Use the two return form of [type assertion](https://golang.org/ref/spec#Type_assertions)

```golang
c, ok := obj.(NewType)
if !ok {
  panic("haha, just kidding")
}
```

evankanderson commented on contrib/gcppubsub/pkg/controller/channel/controller.go:
```diff
@@ -86,3 +105,13 @@ func ProvideController(defaultGcpProject string, defaultSecret *corev1.ObjectRef
+func shouldReconcile(ctx context.Context, obj eventingreconciler.ReconciledResource, r record.EventRecorder) bool {
+	c := obj.(*eventingv1alpha1.Channel) // TODO: Is there a way to do this like try catch - to catch panicking?
+	if c.Spec.Provisioner != nil {
+		return ccpcontroller.IsControlled(c.Spec.Provisioner)
```
Does `IsControlled` not accept a `nil` argument?

akashrv replied:
It does. Will change

evankanderson reviewed (commented):
A few comments on the interface, but it's not clear if this is intended to be a generic reconciler or a Channel-specific reconciler.

mattmoor-sockpuppet committed:
golang format tools (#911)

Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -type f -name '*.go' -print))`
  `goimports -w $(find -name '*.go' | grep -v vendor)`

Harwayne committed:
Mark the Broker's Ingress and Filter status condidionts failed when they have failed.

Harwayne committed:
Merge pull request #45 from Harwayne/broker-mark-failed

Mark the Broker's Ingress and Filter status condidionts failed when they have failed.

Harwayne committed:
Do not DeepCopy() in Reconcile(), as controller-runtime already did it for us.

Harwayne committed:
Merge pull request #46 from Harwayne/broker-no-deepcopy

Do not DeepCopy() in Reconcile(), as controller-runtime already did it for us

Harwayne committed:
Broker, Trigger, and Namespace Controllers (#788)

* Initial API outline for Broker.

* Metadata is exported

* Initial Trigger API.

* Remove old comments.

* Register the types.

* Initial work on the Broker controller.

* Initial work on the Trigger controller.

* Initial work on the new model and its corresponding broker controller.

* Initial work on the Trigger controller.

* Add simple mains.

* Small fixes, still not working.

* Add the Istio injection annotation.

* Resolve subscriber in the Trigger controller.

* Standardize on 'Any'.

* Make Broker and Trigger generational

* Happy control path.

* TargetPort is 8080

* Watch channels.

* Custom service account for filter (needs trigger watch).

* Increase logging in the Filter to debug level.

* Use the default channel provisioner.

* Adding filtering using k8s label selectors

* Watch namespaces and create a default Broker.

* Updating trigger example with filters

* Broker changes cause the namespace watcher to reconcile.

* Move the Logging package import and format.

* Updating after review comments. Only doing exact header matching.
In a follow up PR, I plan to include matching of expressions. It turns
our that they are somewhat more involved than expected.

* Adding filtering expressions. Currently using LabelSelectors without
validations. Using reflection to set some unexposed fields for now.

* Changes to compile

* moving filter

* Moving cmds to broker

* updating controller

* Moving provider and reconciler to reconciler folder, and merging them into one

* Adding verbs to dispatcher

* Moving back mains to cmd

* Updating config

* Moving resources to reconcilers

* Moving broker folder back where it belongs

* Revert "Merge branch 'broker-new-model' into broker-new"

This reverts commit d567412b062233463335c7feea94be795f7f2e15, reversing
changes made to bd6a8fa9c62460ee696854fb44a898e72e249ae2.

* Updating headers and headerExpressions to attributes and attributeExpressions

* Reconcilers notice when mroe things change.

* PR comment.

* Remove redundant watch.

* Unit test scaffold.

* Tests for the namespace reconciler.

* Exact matching for filters without using k8s selectors-based syntax.
The matching can only be done on source and type for now.

* Removing t3 as we don't have set expressions

* Updates after code review

* Attempt to reconcile broker in trigger controller

* Broker controller unit tests.

* Deleting and re-creating subscription object as the backing channel
spec is immutable.

* Adding comment

* Rename Trigger.Spec.Filter.ExactMatch to Trigger.Spec.Filter.SourceAndType. Add unit tests for the filter binary.

* Adding event messages

* Using the broker's namespacedNamed as key to the triggers map.
With this we allow to reconcile only the triggers that belong to the
particular broker that changed

* Adding some tests to trigger

* More UTs

* More UTs

* More UTs

* Namespace reconciler automatically creates the Broker Filter's ServiceAccount and RBAC.

Sadly this doesn't work well because we have such an old version of controller-runtime that the Filter ends up trying to watch _all_ Triggers, not just those in its namespace. And it only gets permission for the Triggers in its own namespace.

* Remove no longer needed label.

* Broker and trigger types UTs

* WIP early E2E test

* Changes after code review. Adding trigger defaults and validation tests

* Cleaner trigger validation

* Adding dummy tests for broker validation...
Should be implemented

* Compiling and moving things around

* Updating test

* More updates

* Waiting for potentially multiple contents.
Removing check for corev1.Service ready.
Removing grant's great design. Just making it simpler for now.

* Compiling

* Fixing compilation

* Fixing compilation errors. Adding AnnotateNamespace function.

* Adding ns

* Adding logs. Changing to lowercase any otherwise the pod name is invalid

* Removing namespace when creating trigger subscriber spec.
Adding wait time constant for default broker creation.

* Checking if all triggers are ready

* Updated logs

* Working

* Adding logs... Still not receiving the events.

* More logs

* Adding build constraint

* Removing unnecessary stuff

* Removing ugly structs

* More logs

* Removing quotes

* More logs

* Adding delay

* Listing triggers in receiver when we create it, so that we don't miss
any message because the client couldn't find the existing trigger.
This is a problem in the in-memory-channel as it doesn't do retries.
Maybe the right solution is to add that functionality there.

* Adding delay to sender pod

* Removing withDelay method and just sleep for a while

* Improve log...

* Updates after code review.

* Adding some more logs and trailing dots.

* Switch import order.

* Updating comments.

* Updating comments.

* Replace the bad errgroup usage with the runnableServer.

* Namespace scoped the Broker Filter's client.

* Fix unit tests.

* Fix yaml

* Setting source to source not type.
Updating comment.

* Switch from annotating the namespace to labeling it, to match Istio.

* General clean up.

* Initial docs for the Broker.

* Fill some of the usage section.

* Add instructions for installing the Broker without using Namespace annotation.

* Create example_{brokers,triggers}.yaml to document how they can be used.

* Fix MD linter issues.

* Fix MD linter issues.

* Fix MD linter issues.

* Minor clean up.

* Clean up some spots the merge didn't catch.

* Fix the bad merge by replacing logger.BaseLogger with logger.FormatLogger.

* Add extra columns when using kubectl get.

* MarkBrokerDoesNotExist

* Rename extra columns.

* Replace the Trigger reconciler's in-memory map with a simple list, utilizing the fact that Controller Runtime already has the results of the List cached.

* Accept v0.1 and v0.2 cloud events. Adding UTs.
Updating initClient as well, removing unnecessary paging.

* Change to resolve.SubscriberSpec().

* Remove restClient as it wasn't actually used.

* Only reconcile the Namespace if the specific resource we care about changes.

* lowercase

* Mark the Broker's Ingress and Filter status condidionts failed when they have failed.

* Do not DeepCopy() in Reconcile(), as controller-runtime already did it for us.

akashrv committed:
Merge branch 'reconcile' of github.com:akashrv/eventing into reconcile

Harwayne commented on pkg/reconciler/reconciler.go:
```diff
@@ -0,0 +1,206 @@
+	return b.r
+}
+
+func (b *ReconcilerBuilderImpl) WithFinalizer(finalizerName string, finalizer FinalizerFunc) ReconcilerBuilder {
```
Not that I can think of why you would have more than one, this implicitly assumes it will be called only once. Maybe `panic` if it is called multiple times?

Actually this seems to be true of almost all the methods on the builder, maybe just document them as such.

Harwayne commented on pkg/reconciler/reconciler.go:
```diff
@@ -0,0 +1,206 @@
+	return nil
+}
+
+// reconcile.Reconciler impl
```
Reconciler -> Reconcile

Harwayne commented on pkg/reconciler/reconciler.go:
```diff
@@ -0,0 +1,206 @@
+	ctx := logging.WithLogger(context.TODO(), r.logger.With(zap.Any("request", request)))
+
+	if r.filter != nil && !r.filter(ctx, obj, r.recorder) {
+		logging.FromContext(ctx).Debug(fmt.Sprintf("Not reconciling %s as ShouldReconcile() returned false", recObjTypeName))
```
Don't need the fmt. What is being reconciled is already in the log's request attribute.

This looks to be true on most (all?) of the logging statements in this method.

Harwayne commented on pkg/reconciler/reconciler.go:
```diff
@@ -0,0 +1,206 @@
+		}
+	}
+
+	updateStatus, result, err := r.ReconcileResource(ctx, obj, r.recorder)
```
My preference would be to calculate `updateStatus` here and not make every  `ReconcileResource` do it.

Harwayne commented on contrib/gcppubsub/pkg/controller/channel/reconcile.go:
```diff
@@ -169,100 +144,70 @@ func (r *reconciler) reconcile(ctx context.Context, c *eventingv1alpha1.Channel)
-	// If we are adding the finalizer for the first time, then ensure that finalizer is persisted
-	// before manipulating GCP PubSub, which will not be automatically garbage collected by K8s if
-	// this Channel is deleted.
-	if addFinalizerResult := util.AddFinalizer(c, finalizerName); addFinalizerResult == util.FinalizerAdded {
```
So one thing that we lose is that with the existing code, the finalizer is only added if a credential and a PubSub client can be created. If for whatever reason they can't (e.g. the secret is badly formed), then the finalizer is not added, so the `Channel` can be deleted easily (without manually removing the finalizer).

We could either accept that we no longer have this nicety. Or we could do something else to indicate we haven't manipulated the real world at all (in fact I think the code is already setup to do this easily).

Harwayne commented on contrib/gcppubsub/pkg/controller/channel/reconcile.go:
```diff
@@ -169,100 +144,70 @@ func (r *reconciler) reconcile(ctx context.Context, c *eventingv1alpha1.Channel)
 
 	c.Status.MarkProvisioned()
-	return false, nil
+	return false, reconcile.Result{}, nil
```
The boolean is wrong. The old code has it meaning 'requeue'. Now it means update the status. It should be `true` here (and probably almost everywhere else too).

akashrv replied:
That is correct. Have already fixed it

Harwayne commented:
> **I strongly recommend not doing this work.** I think if energy is put into reconciling logic improvements, it needs to be in `knative/pkg`. Furthermore, eventing and serving need to match in core implementation so that improvements like this can benefit Knative.

I disagree. I would rather the work gets done on something concrete (e.g. knative/eventing), then moved up to knative/pkg once shown to be working. It is useful to see if this is the pattern would work for knative/serving as well, but I don't think the actual changes to knative/serving needs to be done now.

Overall I like the general idea of this PR. It makes writing all the controllers easier and more uniform.

grantr commented:
If every idea that could possibly apply to multiple projects was initially developed in knative/pkg, that repo would become an experimental dumping ground, not the collection of well-documented, simple, shareable components that we probably want it to be. IMO a package should only be promoted to knative/pkg once it's proven useful in at least one project, though I may be in the minority on that.

Practically, developing this in knative/pkg would be painful. To pull in a new PR, we'd have to upgrade to a new revision of knative/pkg, but that might involve other breakages because knative/pkg has no versioning guarantees and moves as quickly as serving needs it to. It would be much faster and less disruptive to develop in knative/eventing, then consider contributing to knative/pkg as a single PR when it's ready.

akashrv committed:
WIP. Committing to switch branch

n3wscott committed:
Embed a common Status duck type. (#913)

* Update to use ducktype embedded status type.

* Update serving dep to 0.4

* needed new pkg.

* update pkg and serving.

* codegen.

* update using common e2e flags in pkg.

* lint.

* update serving.

* slide the flag values down to eventing.

* adding docs.

* adding why contrain pkg.

* Reran codegen.

* rerun codegen.

brandoncox committed:
Updated the link for the eventing docs so it is pointing to the correct URL (#920)

n3wscott committed:
adding n3wscott to approvers. (#465)

Commit:
Set memory limits (#921)

We're seeing OOM issues at the Node level when we don't set the limit on some pods. This will allow the pod to restart gracefully w/o taking down the entire node.

I'm not stuck on the specific value so we can pick a different one.

See https://github.com/knative/serving/pull/3332 for the serving equivalent.

Signed-off-by: Doug Davis <dug@us.ibm.com>

akashrv committed:
WIP. Committing to switch branch

akashrv committed:
Merge branch 'master' of github.com:knative/eventing into reconcile

akashrv commented:
Update: I'll go ahead and make these changes to all reconcilers in eventing and keep the code in eventing branch for now.
I can create an issue to explore how to move it to knative/pkg and utilize in serving, and handle that later. I did check a reconciler in serving that uses finalizer and didnt find anything common, the reconciler handles it on its own using path. However based on some notes that I read in kubebuilder examples, I think a POST should be good enough

akashrv committed:
New templates to create bug, customer issue and feature request (#919)

* New templates to create bug, customer issue and feature request

* Updates based on PR comments

* Trimmed down to reduce entry threshold to create new issues and bugs

n3wscott committed:
Migrate our usage of pkg/cloudevents to cloudevents sdk. (#924)

* migrate our usage of pkg/cloudevents to cloudevents sdk.

* format.

* format.

* spit out what the test it looking for.

* pointer.

* Fixed receiver.

akashrv committed:
Updated subscription reconciler + relevant UTs
Merge from upstream master

akashrv committed:
Update Channel, and Broker reconciler + UTs

srinivashegde86 committed:
Use common test components defined in pkg/test (#935)

akashrv committed:
Updated namespace reconciler + UTs

akashrv committed:
Updated trigger reconciler + uts

Harwayne committed:
Trigger responses are sent back into the Broker. (#917)

* Initial API outline for Broker.

* Metadata is exported

* Initial Trigger API.

* Remove old comments.

* Register the types.

* Initial work on the Broker controller.

* Initial work on the Trigger controller.

* Initial work on the new model and its corresponding broker controller.

* Initial work on the Trigger controller.

* Add simple mains.

* Small fixes, still not working.

* Add the Istio injection annotation.

* Resolve subscriber in the Trigger controller.

* Standardize on 'Any'.

* Make Broker and Trigger generational

* Happy control path.

* TargetPort is 8080

* Watch channels.

* Custom service account for filter (needs trigger watch).

* Increase logging in the Filter to debug level.

* Use the default channel provisioner.

* Adding filtering using k8s label selectors

* Watch namespaces and create a default Broker.

* Updating trigger example with filters

* Broker changes cause the namespace watcher to reconcile.

* Move the Logging package import and format.

* Updating after review comments. Only doing exact header matching.
In a follow up PR, I plan to include matching of expressions. It turns
our that they are somewhat more involved than expected.

* Adding filtering expressions. Currently using LabelSelectors without
validations. Using reflection to set some unexposed fields for now.

* Changes to compile

* moving filter

* Moving cmds to broker

* updating controller

* Moving provider and reconciler to reconciler folder, and merging them into one

* Adding verbs to dispatcher

* Moving back mains to cmd

* Updating config

* Moving resources to reconcilers

* Moving broker folder back where it belongs

* Revert "Merge branch 'broker-new-model' into broker-new"

This reverts commit d567412b062233463335c7feea94be795f7f2e15, reversing
changes made to bd6a8fa9c62460ee696854fb44a898e72e249ae2.

* Updating headers and headerExpressions to attributes and attributeExpressions

* Reconcilers notice when mroe things change.

* PR comment.

* Remove redundant watch.

* Unit test scaffold.

* Tests for the namespace reconciler.

* Exact matching for filters without using k8s selectors-based syntax.
The matching can only be done on source and type for now.

* Removing t3 as we don't have set expressions

* Updates after code review

* Attempt to reconcile broker in trigger controller

* Broker controller unit tests.

* Deleting and re-creating subscription object as the backing channel
spec is immutable.

* Adding comment

* Rename Trigger.Spec.Filter.ExactMatch to Trigger.Spec.Filter.SourceAndType. Add unit tests for the filter binary.

* Adding event messages

* Using the broker's namespacedNamed as key to the triggers map.
With this we allow to reconcile only the triggers that belong to the
particular broker that changed

* Adding some tests to trigger

* More UTs

* More UTs

* More UTs

* Namespace reconciler automatically creates the Broker Filter's ServiceAccount and RBAC.

Sadly this doesn't work well because we have such an old version of controller-runtime that the Filter ends up trying to watch _all_ Triggers, not just those in its namespace. And it only gets permission for the Triggers in its own namespace.

* Remove no longer needed label.

* Broker and trigger types UTs

* WIP early E2E test

* Changes after code review. Adding trigger defaults and validation tests

* Cleaner trigger validation

* Adding dummy tests for broker validation...
Should be implemented

* Compiling and moving things around

* Updating test

* More updates

* Waiting for potentially multiple contents.
Removing check for corev1.Service ready.
Removing grant's great design. Just making it simpler for now.

* Compiling

* Fixing compilation

* Fixing compilation errors. Adding AnnotateNamespace function.

* Adding ns

* Adding logs. Changing to lowercase any otherwise the pod name is invalid

* Removing namespace when creating trigger subscriber spec.
Adding wait time constant for default broker creation.

* Checking if all triggers are ready

* Updated logs

* Working

* Adding logs... Still not receiving the events.

* More logs

* Adding build constraint

* Removing unnecessary stuff

* Removing ugly structs

* More logs

* Removing quotes

* More logs

* Adding delay

* Listing triggers in receiver when we create it, so that we don't miss
any message because the client couldn't find the existing trigger.
This is a problem in the in-memory-channel as it doesn't do retries.
Maybe the right solution is to add that functionality there.

* Adding delay to sender pod

* Removing withDelay method and just sleep for a while

* Improve log...

* Updates after code review.

* Adding some more logs and trailing dots.

* Switch import order.

* Updating comments.

* Updating comments.

* Replace the bad errgroup usage with the runnableServer.

* Namespace scoped the Broker Filter's client.

* Fix unit tests.

* Fix yaml

* Setting source to source not type.
Updating comment.

* Switch from annotating the namespace to labeling it, to match Istio.

* General clean up.

* Create an ingress channel for Brokers.

* adding cloudevents/sdk-go

* Fix unit tests.

* Delete old recevier.

* Change assumption about response to only being a CloudEvent if it can be parsed as such.

* Clean up from the merge.

* More merge clean up.

* Update cloudevents/sdk-go to 0.4.0.

* Move to the newer, idiomatic CloudEvents library useage.

* tctx.URI is actually just the path...

* Partial update to the unit tests.

* Unit tests working.

* Increase broker_types.go code coverage.

* hack/update-deps.sh

* Increase receiver.go code coverage.

* Increase trigger.go code coverage.

* Constraint, not override.

* broker.go unit tests.

* Better receiver shutdown logic.

akashrv committed:
Update in-memory-channel reconciler + uts

akashrv committed:
updated in-memory CCP reconcilers + uts

dubee committed:
Fix broken links to docs repo (#941)

brandoncox committed:
updated the link to WORKING-GROUPS.md in the README.md of the Eventing repo to point to a working URL (#945)

Harwayne committed:
Tracing works with Broker and Trigger (#936)

* Initial API outline for Broker.

* Metadata is exported

* Initial Trigger API.

* Remove old comments.

* Register the types.

* Initial work on the Broker controller.

* Initial work on the Trigger controller.

* Initial work on the new model and its corresponding broker controller.

* Initial work on the Trigger controller.

* Add simple mains.

* Small fixes, still not working.

* Add the Istio injection annotation.

* Resolve subscriber in the Trigger controller.

* Standardize on 'Any'.

* Make Broker and Trigger generational

* Happy control path.

* TargetPort is 8080

* Watch channels.

* Custom service account for filter (needs trigger watch).

* Increase logging in the Filter to debug level.

* Use the default channel provisioner.

* Adding filtering using k8s label selectors

* Watch namespaces and create a default Broker.

* Updating trigger example with filters

* Broker changes cause the namespace watcher to reconcile.

* Move the Logging package import and format.

* Updating after review comments. Only doing exact header matching.
In a follow up PR, I plan to include matching of expressions. It turns
our that they are somewhat more involved than expected.

* Adding filtering expressions. Currently using LabelSelectors without
validations. Using reflection to set some unexposed fields for now.

* Changes to compile

* moving filter

* Moving cmds to broker

* updating controller

* Moving provider and reconciler to reconciler folder, and merging them into one

* Adding verbs to dispatcher

* Moving back mains to cmd

* Updating config

* Moving resources to reconcilers

* Moving broker folder back where it belongs

* Revert "Merge branch 'broker-new-model' into broker-new"

This reverts commit d567412b062233463335c7feea94be795f7f2e15, reversing
changes made to bd6a8fa9c62460ee696854fb44a898e72e249ae2.

* Updating headers and headerExpressions to attributes and attributeExpressions

* Reconcilers notice when mroe things change.

* PR comment.

* Remove redundant watch.

* Unit test scaffold.

* Tests for the namespace reconciler.

* Exact matching for filters without using k8s selectors-based syntax.
The matching can only be done on source and type for now.

* Removing t3 as we don't have set expressions

* Updates after code review

* Attempt to reconcile broker in trigger controller

* Broker controller unit tests.

* Deleting and re-creating subscription object as the backing channel
spec is immutable.

* Adding comment

* Rename Trigger.Spec.Filter.ExactMatch to Trigger.Spec.Filter.SourceAndType. Add unit tests for the filter binary.

* Adding event messages

* Using the broker's namespacedNamed as key to the triggers map.
With this we allow to reconcile only the triggers that belong to the
particular broker that changed

* Adding some tests to trigger

* More UTs

* More UTs

* More UTs

* Namespace reconciler automatically creates the Broker Filter's ServiceAccount and RBAC.

Sadly this doesn't work well because we have such an old version of controller-runtime that the Filter ends up trying to watch _all_ Triggers, not just those in its namespace. And it only gets permission for the Triggers in its own namespace.

* Remove no longer needed label.

* Broker and trigger types UTs

* WIP early E2E test

* Changes after code review. Adding trigger defaults and validation tests

* Cleaner trigger validation

* Adding dummy tests for broker validation...
Should be implemented

* Compiling and moving things around

* Updating test

* More updates

* Waiting for potentially multiple contents.
Removing check for corev1.Service ready.
Removing grant's great design. Just making it simpler for now.

* Compiling

* Fixing compilation

* Fixing compilation errors. Adding AnnotateNamespace function.

* Adding ns

* Adding logs. Changing to lowercase any otherwise the pod name is invalid

* Removing namespace when creating trigger subscriber spec.
Adding wait time constant for default broker creation.

* Checking if all triggers are ready

* Updated logs

* Working

* Adding logs... Still not receiving the events.

* More logs

* Adding build constraint

* Removing unnecessary stuff

* Removing ugly structs

* More logs

* Removing quotes

* More logs

* Adding delay

* Listing triggers in receiver when we create it, so that we don't miss
any message because the client couldn't find the existing trigger.
This is a problem in the in-memory-channel as it doesn't do retries.
Maybe the right solution is to add that functionality there.

* Adding delay to sender pod

* Removing withDelay method and just sleep for a while

* Improve log...

* Updates after code review.

* Adding some more logs and trailing dots.

* Switch import order.

* Updating comments.

* Updating comments.

* Replace the bad errgroup usage with the runnableServer.

* Namespace scoped the Broker Filter's client.

* Fix unit tests.

* Fix yaml

* Setting source to source not type.
Updating comment.

* Switch from annotating the namespace to labeling it, to match Istio.

* General clean up.

* Create an ingress channel for Brokers.

* adding cloudevents/sdk-go

* Fix unit tests.

* Delete old recevier.

* Change assumption about response to only being a CloudEvent if it can be parsed as such.

* Clean up from the merge.

* More merge clean up.

* Update cloudevents/sdk-go to 0.4.0.

* Move to the newer, idiomatic CloudEvents library useage.

* tctx.URI is actually just the path...

* Partial update to the unit tests.

* Unit tests working.

* Increase broker_types.go code coverage.

* hack/update-deps.sh

* Increase receiver.go code coverage.

* Increase trigger.go code coverage.

* Constraint, not override.

* broker.go unit tests.

* Better receiver shutdown logic.

* cloudevents/sdk-go 0.4.1

* Forward tracing headers.

* Update comment.

* Relearning why message_dispatcher was written the way it was.

* Add a TODO for where to make the change for response tracing too.

Harwayne committed:
Switch the Broker's Ingress to use CloudEvents SDK (#933)

* Initial API outline for Broker.

* Metadata is exported

* Initial Trigger API.

* Remove old comments.

* Register the types.

* Initial work on the Broker controller.

* Initial work on the Trigger controller.

* Initial work on the new model and its corresponding broker controller.

* Initial work on the Trigger controller.

* Add simple mains.

* Small fixes, still not working.

* Add the Istio injection annotation.

* Resolve subscriber in the Trigger controller.

* Standardize on 'Any'.

* Make Broker and Trigger generational

* Happy control path.

* TargetPort is 8080

* Watch channels.

* Custom service account for filter (needs trigger watch).

* Increase logging in the Filter to debug level.

* Use the default channel provisioner.

* Adding filtering using k8s label selectors

* Watch namespaces and create a default Broker.

* Updating trigger example with filters

* Broker changes cause the namespace watcher to reconcile.

* Move the Logging package import and format.

* Updating after review comments. Only doing exact header matching.
In a follow up PR, I plan to include matching of expressions. It turns
our that they are somewhat more involved than expected.

* Adding filtering expressions. Currently using LabelSelectors without
validations. Using reflection to set some unexposed fields for now.

* Changes to compile

* moving filter

* Moving cmds to broker

* updating controller

* Moving provider and reconciler to reconciler folder, and merging them into one

* Adding verbs to dispatcher

* Moving back mains to cmd

* Updating config

* Moving resources to reconcilers

* Moving broker folder back where it belongs

* Revert "Merge branch 'broker-new-model' into broker-new"

This reverts commit d567412b062233463335c7feea94be795f7f2e15, reversing
changes made to bd6a8fa9c62460ee696854fb44a898e72e249ae2.

* Updating headers and headerExpressions to attributes and attributeExpressions

* Reconcilers notice when mroe things change.

* PR comment.

* Remove redundant watch.

* Unit test scaffold.

* Tests for the namespace reconciler.

* Exact matching for filters without using k8s selectors-based syntax.
The matching can only be done on source and type for now.

* Removing t3 as we don't have set expressions

* Updates after code review

* Attempt to reconcile broker in trigger controller

* Broker controller unit tests.

* Deleting and re-creating subscription object as the backing channel
spec is immutable.

* Adding comment

* Rename Trigger.Spec.Filter.ExactMatch to Trigger.Spec.Filter.SourceAndType. Add unit tests for the filter binary.

* Adding event messages

* Using the broker's namespacedNamed as key to the triggers map.
With this we allow to reconcile only the triggers that belong to the
particular broker that changed

* Adding some tests to trigger

* More UTs

* More UTs

* More UTs

* Namespace reconciler automatically creates the Broker Filter's ServiceAccount and RBAC.

Sadly this doesn't work well because we have such an old version of controller-runtime that the Filter ends up trying to watch _all_ Triggers, not just those in its namespace. And it only gets permission for the Triggers in its own namespace.

* Remove no longer needed label.

* Broker and trigger types UTs

* WIP early E2E test

* Changes after code review. Adding trigger defaults and validation tests

* Cleaner trigger validation

* Adding dummy tests for broker validation...
Should be implemented

* Compiling and moving things around

* Updating test

* More updates

* Waiting for potentially multiple contents.
Removing check for corev1.Service ready.
Removing grant's great design. Just making it simpler for now.

* Compiling

* Fixing compilation

* Fixing compilation errors. Adding AnnotateNamespace function.

* Adding ns

* Adding logs. Changing to lowercase any otherwise the pod name is invalid

* Removing namespace when creating trigger subscriber spec.
Adding wait time constant for default broker creation.

* Checking if all triggers are ready

* Updated logs

* Working

* Adding logs... Still not receiving the events.

* More logs

* Adding build constraint

* Removing unnecessary stuff

* Removing ugly structs

* More logs

* Removing quotes

* More logs

* Adding delay

* Listing triggers in receiver when we create it, so that we don't miss
any message because the client couldn't find the existing trigger.
This is a problem in the in-memory-channel as it doesn't do retries.
Maybe the right solution is to add that functionality there.

* Adding delay to sender pod

* Removing withDelay method and just sleep for a while

* Improve log...

* Updates after code review.

* Adding some more logs and trailing dots.

* Switch import order.

* Updating comments.

* Updating comments.

* Replace the bad errgroup usage with the runnableServer.

* Namespace scoped the Broker Filter's client.

* Fix unit tests.

* Fix yaml

* Setting source to source not type.
Updating comment.

* Switch from annotating the namespace to labeling it, to match Istio.

* General clean up.

* Create an ingress channel for Brokers.

* adding cloudevents/sdk-go

* Fix unit tests.

* Delete old recevier.

* Change assumption about response to only being a CloudEvent if it can be parsed as such.

* Clean up from the merge.

* More merge clean up.

* Update cloudevents/sdk-go to 0.4.0.

* Move to the newer, idiomatic CloudEvents library useage.

* tctx.URI is actually just the path...

* Partial update to the unit tests.

* Unit tests working.

* Increase broker_types.go code coverage.

* hack/update-deps.sh

* Increase receiver.go code coverage.

* Increase trigger.go code coverage.

* Constraint, not override.

* broker.go unit tests.

* Better receiver shutdown logic.

* cloudevents/sdk-go 0.4.1

* Forward tracing headers.

* Update comment.

* Switch the ingress point from message receiver and message dispatcher to the CloudEvents SDK.

* Improve the shutdown logic.

* CHANNEL is the host, not a full URI.

* Ingress copies tracing headers.

* Actually use sendingCTX everywhere.

* Add unit test for the pass through headers.

Commit:
User can specify webhook deployment name (#942)

akashrv committed:
updated GCPPubSub reconciler + uts

imjasonh committed:
Point to new ko repo (#947)

n3wscott committed:
Add context from webhook invocation. (#932)

* Stage work to add context from webhook invocation.

* updating deps.

* Update deps.

* Update lock on build.

* Updated serving and build.

* Ask for build directly.

akashrv committed:
updated kafka channel reconciler + uts

akashrv committed:
update kafka ClusterChannelProvisioner reconciler + uts

grantr committed:
Fix some broker vet and lint issues (#952)

* call context cancel func properly

Cancel must be called in all return paths. I think that was the intent
here, but ctx.Done() was accidentally called instead of cancel().

* Move New handler func to main

golint was complaining about returning a pointer to an unexported type.
This function seemed unnecessary so I just removed it.

* rename ceHttp to ceHTTP

golint complains about lowercased Http.

akashrv committed:
Update natss channel reconciler + uts

akashrv committed:
update natss ccp reconcilers + uts

Harwayne committed:
Trigger Reply Tracing (#949)

* CloudEvents SDK 0.4.2.

* Send pass through headers on the Trigger's reply.

* CloudEvents SDK 0.4.3.

* Add headers directly.

samodell committed:
Updating links to docs repo (#954)

akashrv committed:
Wrote UTs for generic reconciler

n3wscott committed:
Adding broker/trigger to the spec. (#943)

* Adding broker/trigger to the spec.

* Update based on feedback.

* Trigger.BrokerExists is now Broker Exists and ready.

* First pass define broker/trigger.

* lint.

* adding overview image.

* use svg.:

* Updated after Evan's comments.

* Adding a description of Broker/Trigger overview.

akashrv committed:
Fixed linter issues in reconciler.go

mattmoor-sockpuppet committed:
Format markdown (#962)

Produced via: `prettier --write --prose-wrap=always $(find -name '*.md' | grep -v vendor | grep -v .github)`

mattmoor-sockpuppet committed:
golang format tools (#965)

Produced via:
  `gofmt -s -w $(find -path './vendor' -prune -o -type f -name '*.go' -print))`
  `goimports -w $(find -name '*.go' | grep -v vendor)`

akashrv committed:
Merge branch 'master' of github.com:knative/eventing into reconcile

akashrv committed:
Fixed UTs after merge from upstream master

akashrv committed:
Fixed soem typos

n3wscott commented:
Please don't work on this. We need to move to `knative/pkg/controller`
